README.md
# mahler-wasm

WASM compiled utilities for [Mahler](https://github.com/balena-io-modules/mahler).

This module uses [wasm-pack](https://rustwasm.github.io/wasm-pack/) to expose some essential 
functions from Rust to make them accessible in Node.js.

## Provided functions

- `diff`. Create a [JSON Patch](https://datatracker.ietf.org/doc/html/rfc6902/) to transform between two JSON serializable objects. This function just exposes the functionality from the [json-patch crate](https://crates.io/crates/json-patch).
- `patch`. Apply a [JSON Patch](https://datatracker.ietf.org/doc/html/rfc6902/) to a JSON serializable object. This function just exposes the functionality from the [json-patch crate](https://crates.io/crates/json-patch).

Example

```typescript
import { diff, patch } from "mahler-wasm";

const left = {
  "title": "Goodbye!",
  "author" : {
    "givenName" : "John",
    "familyName" : "Doe"
  },
  "tags":[ "example", "sample" ],
  "content": "This will be unchanged"
};

const right = {
  "title": "Hello!",
  "author" : {
    "givenName" : "John"
  },
  "tags": [ "example" ],
  "content": "This will be unchanged",
  "phoneNumber": "+01-123-456-7890"
};

const p = diff(left, right);
console.log(p);
// [
//   { "op": "remove", "path": "/author/familyName" },
//   { "op": "remove", "path": "/tags/1" },
//   { "op": "replace", "path": "/title", "value": "Hello!" },
//   { "op": "add", "path": "/phoneNumber", "value": "+01-123-456-7890" },
// ]
const doc = patch(left, p);
console.log(doc);
// {
//   "title": "Hello!",
//   "author" : {
//     "givenName" : "John"
//   },
//   "tags": [ "example" ],
//   "content": "This will be unchanged",
//   "phoneNumber": "+01-123-456-7890"
// }
```

## Installing

This module does not have any dependencies for installing, to install, you can just use 

```
npm install mahler-wasm
```

## Building mahler-wasm

To build the module, these are the dependencies

- [Rust](https://www.rust-lang.org) 
- The [wasm32-wasi](https://bytecodealliance.github.io/cargo-wasi/steps.html) target
- [wasm-pack](https://rustwasm.github.io/wasm-pack/)

Before building, you need to install the repository dependencies

```
# Pull the repository if you haven't done it yet
git clone https://github.com/balena-io-modules/mahler-wasm.git && cd mahler-wasm

# Install dependencies
npm install
```

Assuming you have all dependencies installed, you can build using

```
npm run build
```

And use the following to run tests

```
npm run test
```

## License

Licensed under [Apache License, Version 2.0](LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v0.1.1
## (2023-08-16)

* Configure minimal node version [Felipe Lalanne]

# v0.1.0
## (2023-08-16)

* Create first version [Felipe Lalanne]


tests/node.rs
#![cfg(target_arch = "wasm32")]

extern crate wasm_bindgen_test;
use wasm_bindgen_test::*;

use gloo_utils::format::JsValueSerdeExt;
use mahler_wasm::{diff, patch, Patch};
use serde_json::{from_value, json, Value};
use wasm_bindgen::JsValue;

#[wasm_bindgen_test]
fn lib_test() {
    let left = json!({
      "title": "Goodbye!",
      "author" : {
        "givenName" : "John",
        "familyName" : "Doe"
      },
      "tags":[ "example", "sample" ],
      "content": "This will be unchanged"
    });
    let left = JsValue::from_serde(&left).unwrap();

    let right = json!({
      "title": "Hello!",
      "author" : {
        "givenName" : "John"
      },
      "tags": [ "example" ],
      "content": "This will be unchanged",
      "phoneNumber": "+01-123-456-7890"
    });
    let right = JsValue::from_serde(&right).unwrap();

    let p = diff(&left, &right)
        .or_else(|_| Err("Failed to find patch!"))
        .unwrap();
    assert_eq!(
        p,
        from_value::<Patch>(json!([
          { "op": "remove", "path": "/author/familyName" },
          { "op": "remove", "path": "/tags/1" },
          { "op": "replace", "path": "/title", "value": "Hello!" },
          { "op": "add", "path": "/phoneNumber", "value": "+01-123-456-7890" },
        ]))
        .unwrap()
    );

    let obj = patch(&left, p)
        .or_else(|_| Err("Failed to apply patch!"))
        .unwrap();

    assert_eq!(
        obj.into_serde::<Value>().unwrap(),
        json!({
          "title": "Hello!",
          "author" : {
            "givenName" : "John"
          },
          "tags": [ "example" ],
          "content": "This will be unchanged",
          "phoneNumber": "+01-123-456-7890"
        })
    );
}


src/lib.rs
use gloo_utils::format::JsValueSerdeExt;
use json_patch::{diff as json_diff, patch as json_patch, PatchOperation as Operation};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use tsify::Tsify;

use wasm_bindgen::prelude::*;

#[wasm_bindgen(typescript_custom_section)]
const TS_APPEND_CONTENT: &'static str = r#"
export type Operation =
    | {op: "add", path: string, value: any}
    | {op: "remove", path: string}
    | {op: "replace", path: string, value: any}
    | {op: "move", from: string, path: string}
    | {op: "copy", from: string, path: string}
    | {op: "test", path: string, value: any};
"#;

#[derive(Tsify, Serialize, Deserialize, Debug, PartialEq, Eq)]
#[tsify(into_wasm_abi, from_wasm_abi)]
// We take advantage that the PatchOperation enum in json-patch
// is json serializable to use it as part of the Patch type.
// If it wasn't we would have to re-define the Operation enum and transform
// the json-patch operations into our own.
pub struct Patch(Vec<Operation>);

#[wasm_bindgen]
/// Create a JSON patch to transform between two JSON serializable
/// objects
///
/// A [JSON Patch](https://datatracker.ietf.org/doc/html/rfc6902/) document
/// is a JSON document that represents an array of objects.
/// Each object represents a single operation to be applied to the target JSON document.
///
/// This function uses the Rust [json-patch crate](https://crates.io/crates/json-patch),
/// compiled to WASM, to efficiently calculate the differences between two objects.
///
/// The resulting object can be used with the `patch` function to apply on
/// a new object.
///
/// # Arguments
///
/// * `left` - Source object. It needs to be a JSON serializable object
/// * `right` - Target object. It needs to be a JSON serializable object
///
/// # Examples
/// ```typescript
/// import { diff, patch } from "mahler-wasm";
///
/// const left = {
///   "title": "Goodbye!",
///   "author" : {
///     "givenName" : "John",
///     "familyName" : "Doe"
///   },
///   "tags":[ "example", "sample" ],
///   "content": "This will be unchanged"
/// };
///
/// const right = {
///   "title": "Hello!",
///   "author" : {
///     "givenName" : "John"
///   },
///   "tags": [ "example" ],
///   "content": "This will be unchanged",
///   "phoneNumber": "+01-123-456-7890"
/// };
///
/// const p = diff(left, right);
/// console.log(p);
/// // [
/// //   { "op": "remove", "path": "/author/familyName" },
/// //   { "op": "remove", "path": "/tags/1" },
/// //   { "op": "replace", "path": "/title", "value": "Hello!" },
/// //   { "op": "add", "path": "/phoneNumber", "value": "+01-123-456-7890" },
/// // ]
/// const doc = patch(left, p);
/// console.log(doc);
/// // {
/// //   "title": "Hello!",
/// //   "author" : {
/// //     "givenName" : "John"
/// //   },
/// //   "tags": [ "example" ],
/// //   "content": "This will be unchanged",
/// //   "phoneNumber": "+01-123-456-7890"
/// // }
/// ```
///
pub fn diff(left: &JsValue, right: &JsValue) -> Result<Patch, JsError> {
    let left: Value = left.into_serde()?;
    let right: Value = right.into_serde()?;

    let json_patch::Patch(ops) = json_diff(&left, &right);

    Ok(Patch(ops))
}

#[wasm_bindgen]
/// Apply a JSON patch to a JSON serializable object
///
/// JSON patch 6902
/// A [JSON Patch](https://datatracker.ietf.org/doc/html/rfc6902/) document
/// is a JSON document that represents an array of objects.
/// Each object represents a single operation to be applied to the target JSON document.
///
/// This function uses the Rust [json-patch crate](https://crates.io/crates/json-patch),
/// compiled to WASM, to efficiently apply a patch to an object.
///
/// # Arguments
/// * `obj` - Source object. It needs to be a JSON serializable object
/// * `patch` - Patch to apply.
///
pub fn patch(obj: &JsValue, patch: Patch) -> Result<JsValue, JsError> {
    let mut obj: Value = obj.into_serde()?;

    // Get ops from input
    let Patch(ops) = patch;

    // Apply the patch
    let patch = json_patch::Patch(ops);
    json_patch(&mut obj, &patch)?;

    JsValue::from_serde(&obj).map_err(JsError::from)
}
